src/encoder.rs

```diff
@@ -1110,23 +1110,28 @@ pub fn encode_tx_block<T: Pixel>(
   cw: &mut ContextWriter,
   w: &mut dyn Writer,
   p: usize,
+  // Offset in the luma plane of the partition enclosing this block.
   tile_partition_bo: TileBlockOffset,
-  bx: usize,
-  by: usize, // tx block position within a partition, unit: tx block number
+  // tx block position within a partition, unit: tx block number
+  bx: usize, by: usize,
+  // Offset in the luma plane where this tx block is colocated. Note that for
+  // a chroma block, this offset might be outside of the current partition.
+  // For example in 4:2:0, four 4x4 luma partitions share one 4x4 chroma block,
+  // this block is part of the last 4x4 partition, but its `tx_bo` offset
+  // matches the offset of the first 4x4 partition.
   tx_bo: TileBlockOffset,
   mode: PredictionMode,
   tx_size: TxSize,
   tx_type: TxType,
   bsize: BlockSize,
   po: PlaneOffset,
   skip: bool,
+  qidx: u8,
   ac: &[i16],
   alpha: i16,
   rdo_type: RDOType,
   need_recon_pixel: bool,
 ) -> (bool, ScaledDistortion) {
-  let qidx = get_qidx(fi, ts, cw, tx_bo);
```

Comment:
I have tested your code to check whether the qidx obtained here (before your fix) matches to the qidx passed at line 1129 (after your fix).

Like using either of below:
  //debug_assert!(qidx == qidx2);  // ==> Does not work well in rust native code level debug in my VS Code, so // using below
  if qidx != qidx2 {
    let _test = true;
  }

Then, I am seeing that it consistently mismatches at plane 1 and 2.
The desync starts from the first 1,1 block position (i.e. MI grid) of luma 4x4 block.
Before your fix, instead of passing qidx directly, "cw.bc.blocks.set_segmentation_idx(tile_bo, bsize, sidx);" is used to set (or rememeber) the seg_id at MI grid, but someho chroma blocks fail to retrieve that info, possibly because of using incorrect offset in block context array.

Reply:
The problem is that this didn't use tile_partition_bo, see the commit message for details.

Reply:
> The problem is that this didn't use tile_partition_bo, see the commit message for details.

As you exactly stated, I realized that using the existing parameter 'tile_partition_bo' to get qidx, i.e. doing: let qidx = get_qidx(fi, ts, cw, tile_partition_bo); fixes it as well.
However, I believe you didn't like fetching seg_id from block context and wanted passing qidx (since it is already available from caller functions, write_tx_blocks() and write_tx_tree()).

As in 
https://github.com/ycho/rav1e/commit/a8f10afdc9ab9221385906b8a8d205d11202a358